robyn/openapi.py

```diff
@@ -0,0 +1,263 @@
+
+        :param subrouter_openapi: the OpenAPI object of the current subrouter
+        """
+        paths = subrouter_openapi.openapi_spec["paths"]
+
+        for path in paths:
+            self.openapi_spec["paths"][path] = paths[path]
+
+    def get_path_obj(self, endpoint: str, summary: str, tags: list, query_params: TypedDict = None):
+        """
+        get the "path" openapi object according to spec
+
+        :param endpoint: the endpoint to be added
+        :param summary: short summary of the endpoint (to be fetched from the endpoint defenition by default)
+        :param tags: tags -- for grouping of endpoints
+        :param query_params: query params for the function
+
+        :return: the "path" openapi object according to spec
+        """
+        # for converting path param from `:param` to `{param}`
+        # initialized with endpoint for handling endpoints without path params
+        endpoint_with_path_params_wrapped_in_braces = endpoint
+
+        path_params = endpoint.split(":")
+        openapi_parameter_object = []
+
+        if len(path_params) > 1:
+            path = path_params[0]
+
+            endpoint_with_path_params_wrapped_in_braces = path.removesuffix("/")
+
+            for param in path_params[1:]:
+                param_name = param[:-1] if param.endswith("/") else param
+
+                openapi_parameter_object.append(
+                    {
+                        "name": param_name,
+                        "in": "path",
+                        "required": True,
+                        "schema": {"type": "string"},
+                    }
+                )
+                endpoint_with_path_params_wrapped_in_braces += "/{" + param_name + "}"
+
+        if query_params:
+            for query_param in query_params.__annotations__:
+                # ugly hack!
+                # returns "<class 'int'>" -- this line strips out the type a.k.a int from it
+                param_type = re.findall("'[a-z]+'", str(query_params.__annotations__[query_param]))[0].replace("'", "")
+
+                openapi_parameter_object.append(
+                    {
+                        "name": query_param,
+                        "in": "query",
+                        "required": False,
+                        "schema": {"type": param_type},
+                    }
+                )
+
+        return endpoint_with_path_params_wrapped_in_braces, {
+            "summary": summary,
+            "tags": tags,
+            "parameters": openapi_parameter_object,
+            "responses": {
+                "200": {
+                    "description": "Successful Response",
+                }
+            },
+        }
+
+    def docs_handler(self):
+        """
+        get the swagger html page
+        side effect: this function also dumps the openapi json file
+        @return: the swagger html page
+        """
+        json.dumps(self.openapi_spec)
```

Comment:
I believe we should check if there is an openapi file already present in that location. And json.dumps should be a part of a different function, which is being called in the `docs_handler` function.

Reply:
> I believe we should check if there is an openapi file already present in that location. 

this function only gets called whilst running the application. and, IMO the file should be overwritten since the app & the endpoints might have changed.

> And json.dumps should be a part of a different function, which is being called in the docs_handler function.

sure, done.